File: python/leetcode/24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head is None:
            return None
        if head.next is None:
            return head

        cur = head.next
        head.next = cur.next
        cur.next = head
        head = cur

        head.next.next = self.swapPairs(head.next.next)
        return head

File: javascript/leetcode/24. Swap Nodes in Pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  if (!head) return null;
  if (!head.next) return head;

  let cur = head.next;
  head.next = cur.next;
  cur.next = head;
  head = cur;

  head.next.next = swapPairs(head.next.next);

  return head;
};
